File: Bots/Phase3Bot.js
//IMPORTS
const config = require("./config")


//GLOBAL VARS
const Web3 = require("web3");
const provider = new Web3.providers.HttpProvider(config.web3Provider);
const web3 = new Web3(provider);
web3.eth.net.isListening()
.then(() => console.log('web3 is connected'))
.catch(e => console.log('Wow. Something went wrong'));


const casinoContract = new web3.eth.Contract(config.abi,config.smartContractAddress)


web3.eth.accounts.wallet.add('0x' + config.casinoPrivateKey )


var accounts = web3.eth.accounts.wallet
var account = accounts["0"]
var myAddress = account["address"]


// getActiveGames -> Retrieves all of the games that are in Phase 2


const openCards = () => {
    var array = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51];
    var currentIndex = array.length, temporaryValue, randomIndex;
  
    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
  
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
  
      // And swap it with the current element.
      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }
  
    return array;
  }
  

const getPhase3Games = async () => {
   const temp = await casinoContract.methods.getPhase3Games().call({from: myAddress})
   activeGames = []
   for (i = 0; i < temp.length; i++){
       if (temp[i] != '0x0000000000000000000000000000000000000000'){
           activeGames.push(temp[i])
       }
   }
   return activeGames
}


const phase3Response =  async (activeGames) => {
    if (activeGames.length > 0){
        console.log("Detected the following games")
        console.log(activeGames)
        console.log()
    }
    for (i = 0; i < activeGames.length; i++){
        activeGame = activeGames[i]
        console.log("Distribute cards for user and dealer for address:",activeGame)

        const params = {
            name: 'Casino_get_deck',
            type: 'function',
            inputs: [
                {
                    type: 'address',
                    name: activeGame
                }
            ]
        }
    
        const gasPrice = await web3.eth.getGasPrice();
        const gasEstimate = await casinoContract.methods.distribute(activeGame).estimateGas({ from: myAddress, to: activeGame  });
        
        const response = await casinoContract.methods.distribute(activeGame).send({from: myAddress, to: activeGame, gasPrice: gasPrice, gas: gasEstimate })
        console.log("Done!")
        console.log(response)
        console.log()
    }
}


const main = async () => {



    const phase3Bot =  async () => {
        var activeGames =  await getPhase3Games()
        console.log(activeGames)
        await phase3Response(activeGames)
        setTimeout(phase3Bot,5000)
    }
    phase3Bot()


    
}

main()

File: Bots/Phase3Bot.py
from web3 import Web3
import config
import math
import random
import time
import asyncio

w3 = Web3(Web3.HTTPProvider(config.infuraProvider))
accounts = w3.eth.account.privateKeyToAccount(config.casinoPrivateKey)
account = accounts.address
smartContractAddress = Web3.toChecksumAddress(config.smartContractAddress)
casinoContract = w3.eth.contract(address=smartContractAddress, abi=config.abi)

def getPhase3Games():
    activeGames = []
    temp = casinoContract.functions.getPhase3Games().call({'from': account})
    for i in range(len(temp)):
        if (temp[i] != '0x0000000000000000000000000000000000000000'):
            activeGames.append(temp[i])
    return activeGames

async def asyncFunction(delay,func):
       await asyncio.sleep(delay)
       return func


async def phase3Response(activeGame,nonce):

       try:
              print("\tProcessing game for {}".format(str(activeGame)))
              print("\tDistributing cards for user and dealer for address:", activeGame)
              transaction = await asyncFunction(casinoContract.functions.distribute(activeGame).buildTransaction(
                     {'nonce': nonce, 'from': account, 'to': activeGame}))
              signed_tx = await asyncFunction(w3.eth.account.sign_transaction(transaction, config.casinoPrivateKey))
              response = await asyncFunction(w3.eth.send_raw_transaction(signed_tx.rawTransaction))
              print(response)
              print("\tDone")
              print()
       except Exception as e: 
              print(str(e))


nonce = w3.eth.getTransactionCount(account)

while True:
       print("- - - - - - - - - -")
       print("Scanning games...")
       print()
       phase3Games = getPhase3Games()
       print(phase3Games)
       if len(phase3Games) > 0:
              print("Games in phase 3 detected:")
              for game in phase3Games:
                     print("\t{}".formate(str(game)))
              print()
              for game in phase3Games:
                     asyncio.run(phase3Response(game,nonce))
                     nonce += 1
              print()
       time.sleep(5)
